substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -2160,164 +2166,54 @@ public NativeImageDebugLocationInfo split(int stackDecrement, int adjustment) {
-                fregs = AARCH64_FREG;
-                stackOffset = AARCH64_STACK_OFFSET;
-            } else {
-                if (os == OS.LINUX) {
-                    gpregs = AMD64_GPREG_LINUX;
-                    fregs = AMD64_FREG_LINUX;
-                } else {
-                    gpregs = AMD64_GPREG_WINDOWS;
-                    fregs = AMD64_FREG_WINDOWS;
-                }
-                stackOffset = AMD64_STACK_OFFSET;
-            }
-            nextGPRegIdx = 0;
-            nextFPregIdx = 0;
-            nextStackIdx = 0;
-            stackParamCount = computeStackCount(method);
-        }
-
-        public NativeImageDebugLocalValue nextLocation(JavaKind kind) {
-            switch (kind) {
-                case Float:
-                case Double:
-                    return nextFloatingLocation();
-                case Void:
-                case Illegal:
-                    assert false : "unexpected parameter kind in next location request";
-                    return null;
-                default:
-                    return nextIntegerLocation();
-            }
+        NativeImageDebugLocalValue thisLocation() {
+            assert !hostedMethod.isStatic();
+            return unpack(callingConvention.getArgument(0));
         }
 
-        public NativeImageDebugLocalValue nextFloatingLocation() {
-            if (nextFPregIdx < fregs.length) {
-                return NativeImageDebugRegisterValue.create(fregs[nextFPregIdx++].number);
-            } else {
-                return nextStackLocation();
+        NativeImageDebugLocalValue paramLocation(int paramIdx) {
+            assert paramIdx < hostedMethod.getSignature().getParameterCount(false);
+            int idx = paramIdx;
+            if (!hostedMethod.isStatic()) {
+                idx++;
             }
+            return unpack(callingConvention.getArgument(idx));
         }
 
-        public NativeImageDebugLocalValue nextIntegerLocation() {
-            if (nextGPRegIdx < gpregs.length) {
-                return NativeImageDebugRegisterValue.create(gpregs[nextGPRegIdx++].number);
+        private NativeImageDebugLocalValue unpack(AllocatableValue value) {
+            if (value instanceof RegisterValue) {
+                RegisterValue registerValue = (RegisterValue) value;
+                return NativeImageDebugRegisterValue.create(registerValue);
             } else {
-                return nextStackLocation();
+                // call argument must be a stack slot if it is not a register
+                StackSlot stackSlot = (StackSlot) value;
+                this.usesStack = true;
+                // the calling convention provides offsets from the SP relative to the current
+                // frame size. At the point of call the frame may or may not include a return
+                // address depending on the architecture.
+                return NativeImageDebugStackValue.memoizedCreate(stackSlot.getOffset(PRE_EXTEND_FRAME_SIZE));
```

Comment:
nit: I'd use `create` here for consistency.
```suggestion
                return NativeImageDebugStackValue.create(stackSlot, PRE_EXTEND_FRAME_SIZE);
```

Reply:
Done